File: programmers/level2/주식가격.py
def solution(prices):
    answer = [0] * len(prices)
    stack = []
    for time, pr in enumerate(prices):
        while stack and prices[stack[-1]] > pr: #지금 보고 있는 가격보다 더 높은 가격이 과거 가격에 있으면
            #주식 가격이 과거 한 시점의 관찰가격보다 하락함
			temp_time = stack.pop() #높은 가격이 나왔던 시간
            answer[temp_time] = time - temp_time
			#stack의 top만 봐도 상관 없는 이유는, 매 for문마다 검사하기 때문이다.
        stack.append(time)
    while stack:
        temp_time = stack.pop()
        answer[temp_time] = len(prices) - temp_time - 1
    return answer

File: programmers/level2/주식가격.java
import java.util.*;
class Solution {
    public int[] solution(int[] prices) {
        ArrayList<PriceInfo> arr = new ArrayList<>();
        int[] answer = new int[prices.length];
        PriceInfo pi;
        for(int i=0; i<prices.length; i++) {
            int j = arr.size()-1;
            while(j>=0) {
                pi = arr.get(j--);
                if(pi.price > prices[i]) {
                    answer[pi.idx] = i-pi.idx;
                    arr.remove(j+1);
                }
            }
            arr.add(new PriceInfo(i, prices[i]));
        }
        while(arr.size()>0) {
            int idx = arr.get(arr.size()-1).idx;
            answer[idx] = prices.length-idx-1;
            arr.remove(arr.size()-1);
        }
        return answer;
    }
    
    class PriceInfo {
        int idx;
        int price;
        PriceInfo(int idx, int price) {
            this.idx = idx;
            this.price = price;
        }
    }
}